src/java/org/apache/cassandra/journal/ActiveSegment.java

```diff
@@ -261,21 +264,35 @@ boolean shouldFlush()
      * TODO FIXME: calls from outside Flusher + callbacks
      * @return last synced offset
      */
-    synchronized int flush()
+    synchronized int flush(boolean fsync)
     {
         int allocatePosition = this.allocatePosition.get();
         if (lastFlushedOffset >= allocatePosition)
             return lastFlushedOffset;
 
         waitForModifications();
-        flushInternal();
+        if (fsync) flushInternal();
         lastFlushedOffset = allocatePosition;
+        if (fsync)
+            lastFsyncOffsetUpdater.accumulateAndGet(this, allocatePosition, Math::max);
         int syncedOffset = Math.min(allocatePosition, endOfBuffer);
-        syncedOffsets.mark(syncedOffset);
+        syncedOffsets.mark(syncedOffset, fsync);
         flushComplete.signalAll();
         return syncedOffset;
     }
 
+    // provides no ordering guarantees
+    void fsync()
+    {
+        int lastFlushed = lastFlushedOffset;
+        if (lastFsyncOffset >= lastFlushed)
+            return;
+
+        flushInternal();
```

Comment:
Rename to `fsyncInternal()`? flush vs fsync terminology becomes a little confusing here overall, I wonder if there is a better way to deconfuse these.